saleor/graphql/discount/mutations/voucher/voucher_create.py

```diff
@@ -110,27 +143,152 @@ def clean_input(cls, info: ResolveInfo, instance, data, **kwargs):
                     )
                 }
             )
-        cleaned_input = super().clean_input(info, instance, data, **kwargs)
 
-        return cleaned_input
+    @classmethod
+    def _clean_new_codes(cls, data):
+        codes = [code.strip() for code in data.add_codes]
+        duplicated_codes = get_duplicated_values(codes)
+
+        if duplicated_codes:
+            raise ValidationError(
+                {
+                    "codes": ValidationError(
+                        "Duplicated promo codes provided.",
+                        code=DiscountErrorCode.DUPLICATED_INPUT_ITEM.value,
+                        params={"voucher_codes": duplicated_codes},
+                    )
+                }
+            )
+
+        existing_codes = []
+        clean_add_codes = []
+        for code in data.add_codes:
+            cleaned_code = code.strip() if code else None
+
+            if not cleaned_code:
+                cleaned_code = generate_promo_code()
+            elif not is_available_promo_code(cleaned_code):
+                existing_codes.append(cleaned_code)
+
+            clean_add_codes.append(cleaned_code)
+
+        if existing_codes:
+            raise ValidationError(
+                {
+                    "codes": ValidationError(
+                        "Promo code already exists.",
+                        code=DiscountErrorCode.ALREADY_EXISTS.value,
+                        params={"voucher_codes": existing_codes},
+                    )
+                }
+            )
+
+        data["add_codes"] = clean_add_codes
 
     @classmethod
-    def success_response(cls, instance):
-        instance = ChannelContext(node=instance, channel_slug=None)
-        return super().success_response(instance)
+    def clean_codes(cls, data):
+        if data.code != "":
+            validate_one_of_args_is_in_mutation(
+                "code",
+                data.code,
+                "add_codes",
+                data.add_codes,
+                use_camel_case=True,
+            )
+
+        if "code" in data:
+            cls._clean_old_code(data)
+
+        if "add_codes" in data:
+            cls._clean_new_codes(data)
+
+    @classmethod
+    def construct_codes_instances(
+        cls, code, codes_data, cleaned_input, voucher_instance
+    ):
+        if code:
```

Comment:
Are you sure, that the order is correct here? 
I would say that we should use `codes_data` first. If missing, then use it as a fallback `code`. IMHO We should always use deprecated fields at the end.

Reply:
Also, you should create single codes list based on `code`+`add_codes` OR provide the proper description in the input that will tell, when `code` or `add_codes` are ignored.

Reply:
we have `validate_one_of_args_is_in_mutation` do you can't send `code` and `add_codes` together,  so there will always be one list